libs/langchain-community/src/langgraph/checkpointers/vercel_kv.ts

```diff
@@ -0,0 +1,152 @@
+import { VercelKV, createClient } from "@vercel/kv";
+
+import { RunnableConfig } from "@langchain/core/runnables";
+import {
+  BaseCheckpointSaver,
+  Checkpoint,
+  CheckpointMetadata,
+  CheckpointTuple,
+  SerializerProtocol,
+} from "@langchain/langgraph/web";
+
+// snake_case is used to match Python implementation
+interface KVRow {
+  checkpoint: string;
+  metadata: string;
+}
+
+interface KVConfig {
+  url: string;
+  token: string;
+}
+
+export class VercelKVSaver extends BaseCheckpointSaver {
+  private kv: VercelKV;
+
+  constructor(config: KVConfig, serde?: SerializerProtocol<unknown>) {
+    super(serde);
+    this.kv = createClient(config);
+  }
+
+  async getTuple(config: RunnableConfig): Promise<CheckpointTuple | undefined> {
+    const thread_id = config.configurable?.thread_id;
+    const checkpoint_id = config.configurable?.checkpoint_id;
+
+    if (!thread_id) {
+      return undefined;
+    }
+
+    const key = checkpoint_id
+      ? `${thread_id}:${checkpoint_id}`
+      : `${thread_id}:last`;
+
+    const row: KVRow | null = await this.kv.get(key);
+
+    if (!row) {
+      return undefined;
+    }
+
+    const [checkpoint, metadata] = await Promise.all([
+      this.serde.parse(row.checkpoint),
+      this.serde.parse(row.metadata),
+    ]);
+
+    return {
+      checkpoint: checkpoint as Checkpoint,
+      metadata: metadata as CheckpointMetadata,
+      config: checkpoint_id
+        ? config
+        : {
+            configurable: {
+              thread_id,
+              checkpoint_id: (checkpoint as Checkpoint).id,
+            },
+          },
+    };
+  }
+
+  async *list(
+    config: RunnableConfig,
+    limit?: number,
+    before?: RunnableConfig
+  ): AsyncGenerator<CheckpointTuple> {
+    const thread_id = config.configurable?.thread_id;
+
+    // LUA script to get keys excluding those starting with "last"
+    const luaScript = `
+      local prefix = ARGV[1] .. ':'
+      local keys = redis.call('keys', prefix .. '*')
+      local result = {}
+      for _, key in ipairs(keys) do
+        if string.sub(key, string.len(prefix) + 1, string.len(prefix) + 4) ~= 'last' then
```

Comment:
nit: Use `:last` as a suffix

Reply:
I optimized the keys lookup and added the suggested check on `:last`